lib/websocket.js

```diff
@@ -27,64 +74,406 @@ function init (env, ctx, server) {
+      //    }
+      //  }
+      socket.on('dbUpdate', function dbUpdate (data, callback) {
+       console.log(LOG_WS + 'dbUpdate client ID: ', socket.client.id, ' data: ', data);
+       var collection = supportedCollections[data.collection];
+        
+        var check = checkConditions('dbUpdate', data);
+        if (check) {
+         if (callback) {
+            callback( check );
+          }
+          return;
+        }
+        
+        ctx.store.db.collection(collection).update(
+          { '_id': new ObjectID(data._id) },
+          { $set: data.data }
+        );
+          
+        if (callback) {
+          callback( { result: 'success' } );
+        }
+        ctx.bus.emit('data-received');
+      });
+      
+      // dbUpdateUnset message  
+      //  {
+      //    collection: treatments
+      //    _id: 'some mongo record id'
+      //    data: {
+      //      field_1: 1,
+      //      field_2: 1
+      //    }
+      //  }
+      socket.on('dbUpdateUnset', function dbUpdateUnset (data, callback) {
+       console.log(LOG_WS + 'dbUpdateUnset client ID: ', socket.client.id, ' data: ', data);
+       var collection = supportedCollections[data.collection];
+        
+        var check = checkConditions('dbUpdate', data);
+        if (check) {
+         if (callback) {
+            callback( check );
+          }
+          return;
+        }
+        
+        ctx.store.db.collection(collection).update(
+          { '_id': new ObjectID(data._id) },
+          { $unset: data.data }
+        );
+          
+        if (callback) {
+          callback( { result: 'success' } );
+        }
+        ctx.bus.emit('data-received');
+      });
+      
+      // dbAdd message  
+      //  {
+      //    collection: treatments
+      //    data: {
+      //      field_1: new_value,
+      //      field_2: another_value
+      //    }
+      //  }
+      socket.on('dbAdd', function dbAdd (data, callback) {
+       console.log(LOG_WS + 'dbAdd client ID: ', socket.client.id, ' data: ', data);
+        var collection = supportedCollections[data.collection];
+        var maxtimediff = times.mins(3).msecs;
+        
+        var check = checkConditions('dbAdd', data);
+        if (check) {
+         if (callback) {
+            callback( check );
+          }
+          return;
+        }
+        
+        if (data.collection === 'treatments' && !('eventType' in data.data)) {
+          data.data.eventType = '<none>';
+        }
+        if (!('created_at' in data.data)) {
+          data.data.created_at = new Date().toISOString();
+        }
+        
+        // treatments deduping
+        if (data.collection === 'treatments') {
+          var query  = {
+            created_at: data.data.created_at
+            , eventType: data.data.eventType
+          };
+         
+         // try to find exact match
+          ctx.store.db.collection(collection).find(query).toArray(function findResult (err, array) {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [This function's cyclomatic complexity is too high. (10)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/2651207102/issues/source?bid=2452374&fileBranchId=3324571#l263)